walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,54 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> openTimes = new ArrayList<TimeRange>();
+
+    // Edge case when a meeting is longer than a full day.
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()){
+      return openTimes;
+    }
+    // Make a sorted list of the events.
+    ArrayList<Event> eventList = new ArrayList(events);
+    Collections.sort(eventList, (Event e1, Event e2) -> TimeRange.ORDER_BY_START.compare(e1.getWhen(), e2.getWhen()));
+    
+    int dayStart = TimeRange.START_OF_DAY;
+
+    for (Event event: events) {
+      // Ignore the event if no required members are in it.
+      if (Collections.disjoint(event.getAttendees(), request.getAttendees())){
+        continue;
+      }
+
+      if (dayStart < event.getWhen().start()) {
+        // If the meeting can happen before the event starts, add it.
+        if (dayStart + request.getDuration() <= event.getWhen().start()) {
+          openTimes.add(TimeRange.fromStartEnd(dayStart, event.getWhen().start(), false));
+        }
+
+        // Move start pointer to after the event too keep track of an open time.
+        dayStart = event.getWhen().end();
+        continue;
+        
+      } 
+      // If the start pointer is in the middle of the event, move the start pointer to after the event.
+      else if (dayStart < event.getWhen().end()) {
+        dayStart = event.getWhen().end();
```

Comment:
looks like you do this in both the if and else if statement, can we instead pull this out to happen regardless of an `if` and get rid of this else if all together?
e.g.
`if dayStart < when_event_starts && dayStart + event_duration < when_event_starts {
  // add to open times
}
// Move pointer to end of event
dayStart = event.getWhen().end();`

Reply:
I just tried this out and it wouldn't work due to the fact that the else if checks if our start pointer is in the middle of the event so we can update the pointer to just skip tracking that event all together. I should only update the pointer in those two scenarios, other than that the pointer would skip potential open times until the next event.

Reply:
Ah I missed, thanks!